fdbserver/BlobManager.actor.cpp

```diff
@@ -5243,6 +5243,63 @@ ACTOR Future<Void> bgConsistencyCheck(Reference<BlobManagerData> bmData) {
 	}
 }
 
+// Check if we need to truncate mutation logs according to log retention seconds
+ACTOR Future<bool> shouldTruncateMutationLogs(Reference<BlobManagerData> bmData, BackupDescription desc) {
+	wait(desc.resolveVersionTimes(bmData->db));
+
+	if (!desc.minLogBegin.present() || !desc.contiguousLogEnd.present()) {
+		TraceEvent("MissingMutationLogVersions").log();
+		throw blob_restore_missing_logs();
+	}
+
+	Version logStartVersion = desc.minLogBegin.get();
+	auto start = desc.versionTimeMap.find(logStartVersion);
+	if (start == desc.versionTimeMap.end()) {
+		TraceEvent("MissingVersionTimeMap").detail("LogStartVersion", logStartVersion);
+		throw blob_restore_missing_logs();
+	}
+
+	Version logEndVersion = desc.contiguousLogEnd.get();
+	auto end = desc.versionTimeMap.find(logEndVersion);
+	if (end == desc.versionTimeMap.end()) {
+		TraceEvent("MissingVersionTimeMap").detail("LogEndVersion", logEndVersion);
+		throw blob_restore_missing_logs();
+	}
+
+	if (end->second - start->second < SERVER_KNOBS->BLOB_RESTORE_MLOGS_RETENTION_SECS) {
+		TraceEvent("SkipMutationLogTruncation").detail("EndTs", end->second).detail("StartTs", start->second);
+		return false;
+	}
+
+	return true;
+}
+
+ACTOR Future<Void> maybeTruncateMutationLogs(Reference<BlobManagerData> bmData) {
+	try {
+		std::string mlogsUrl = wait(getMutationLogUrl());
+		state Reference<IBackupContainer> bc = IBackupContainer::openContainer(mlogsUrl, {}, {});
+		state BackupDescription desc = wait(bc->describeBackup(true));
+		bool shouldTruncate = wait(shouldTruncateMutationLogs(bmData, desc));
+		if (!shouldTruncate) {
+			return Void();
+		}
+
+		// Truncate half of mutation logs
+		state Version truncVersion = (desc.contiguousLogEnd.get() - desc.minLogBegin.get()) / 2;
```

Comment:
Why truncate to half of the mutation logs?
It feels like either we want to flush granules at the latest version so the amount of mutation logs we have to apply on restore is minimal, or always keep a specific window of mutation logs and flush/truncate at latest-window version?

Reply:
discussed offline about this topic. I'll follow-up to clarify the concept of "restorable window" that you mentioned. thank you!

Reply:
Wouldn't you now want to flush/truncate just to logEndVersion - BLOB_RESTORE_MLOGS_RETENTION_SECS? So that you always keep at least BLOB_RESTORE_MLOGS_RETENTION_SECS of mutation logs?
And you want to manage manifests the same way, right? Otherwise you can't restore back to now() - BLOB_RESTORE_MLOGS_RETENTION_SECS manifest properly?
I don't understand how the half logic applies here once you have the window.

The other problem is that once you did that though, you'd be flushing + truncating every time you saved a manifest, which causes way too many granule flushes.
If we want to keep the granule flushes very infrequent, we'd likely want to add a flush interval, and only flush+truncate at version end-retention_sec if the logs have up to end-retention_sec-flush_interval, or something like that

Reply:
we cannot use version to substract the seconds? Probably you meant to use VERSIONS_PER_SECOND * BLOB_RESTORE_MLOGS_RETENTION_SECS? How accurate is this estimation? 

It won't flushing + truncating every time when dumping manifest. The check is done in shouldTruncateMutationLogs(). Since we truncate half the mutation logs. LogStartVersion should be way newer than  BLOB_RESTORE_MLOGS_RETENTION_SECS(considering it's 14 days).  

I can also truncate to logEndVersion - VERSIONS_PER_SECOND * BLOB_RESTORE_MLOGS_RETENTION_SECS as well. I assume it should be near to half mutation logs as well(considering a 14 days)?